drivers: add imx7d pwm driver

diegosueiro opened the pull request:
This pull request is to add PWM driver for mcimx7_m4 arch and colibri_imx7d_m4 board.

The implementation was tested on the colibri_imx7d_m4 board by compiling and running the samples/basic/blink_led and tests/drivers/pwm/pwm_api/ applications (PWM characteristics were verified using a Saleae Logic Analyzer) using the PWM_1 interface.

diegosueiro committed:
arch: Add support for i.MX PWM

Adds definitions, devicetree entries and clock controller
configurations for PWM peripheral.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

dbkinder reviewed (approved):
+1 for doc changes

MaureenHelm reviewed (changes requested):
Please update `samples/basic/blink_led/sample.yaml` and `tests/drivers/pwm/pwm_api/testcase.yaml` to include the colibri board. This will ensure that we build the driver in CI

diegosueiro commented on ext/hal/nxp/imx/devices/MCIMX7D/clock_freq.c:
```diff
@@ -238,6 +238,52 @@ uint32_t get_uart_clock_freq(UART_Type *base)
+ * Description   : Get clock frequency applys to the PWM module
+ *
+ *END**************************************************************************/
+uint32_t get_pwm_clock_freq(PWM_Type *base)
```
@MaureenHelm,

The clock_freq.c came originally from "FreeRTOS_BSP_1.0.1_iMX7D/examples/imx7d_val_m4/clock_freq.c" and I put it inside "ext/hal/nxp/imx/devices/MCIMX7D/" path following [yours](https://github.com/zephyrproject-rtos/zephyr/pull/6367#discussion_r170664866) and @stanislav-poboril [suggestion](https://github.com/zephyrproject-rtos/zephyr/pull/6367#pullrequestreview-103805824).

What should I do in this case? Move all get_pwm_clock_freq related stuff to "arch/arm/soc/nxp_imx/mcimx7_m4/" folder?

stanislav-poboril replied:
The clock_freq.c should stay in the ext folder, unmodified. The get_pwm_clock_freq() function and related changes should be either moved to soc.c or Zephyr has a mechanism for this which is probably preferred but would require a little more effort - see drivers/clock_control/ folder and the file clock_control_mcux_ccm.c.

diegosueiro replied:
@MaureenHelm and @stanislav-poboril ,

This should be addressed now.

MaureenHelm replied:
Just to be clear, did you write `get_pwm_clock_freq()` yourself following the pattern from clock_freq.c, or did you copy the function from somewhere? I assumed the former in my previous comment.

diegosueiro replied:
@MaureenHelm, 

Yes, I wrote the function by myself following the same pattern from the other functions.

diegosueiro commented:
@MaureenHelm,

I found some issues with the pwm_api test related to asserts in the i.MX UART Driver. This PR depends on the PR #8938.

diegosueiro committed:
samples: Add colibri_imx7d_m4 config in blink_led

Adds the appropriated configuration for colibri_imx7d_m4 board in
blink_led sample app.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
tests: Add colibri_imx7d_m4 config in pwm_api

Adds the appropriated configuration for colibri_imx7d_m4 board in
pwm_api tests.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

stanislav-poboril commented on drivers/pwm/pwm_imx.c:
```diff
@@ -0,0 +1,205 @@
+
+	if (data->period_cycles != period_cycles) {
+		data->period_cycles = period_cycles;
+		PWM_PWMPR_REG(base) = period_cycles;
```
If there are still some data in the sample register FIFO, say 1-3, could this happen:
- We change the period to another value.
- Samples in the FIFO were meant to be output with the previous period value.
- But they will be output with the new period value, it may even be shorter than the sample value.

This is probably not very usual to change the period and I know the hardware does not allow to handle it better, though (unless we maintain our own fifo with both the cycle and period and set the HW FIFO low watermark to 1 and change the period in interrupt at the last moment).

diegosueiro replied:
@stanislav-poboril ,

Maybe in the "if (enable)" block it can wait for a full cycle if there is still 1,2,3 or 4 words of data in the FIFO. I think this way it will only configure a new period cycle if the FIFO is empty.
What do you think?

stanislav-poboril replied:
The best would be to try it. But then it could be perhaps too late to set the new period - there could be delays between stream of output with different cycles, which means FIFO would be emptied and the last cycle would repeat once more before changing the period.

I think the peripheral was created with the following use case in mind where period is set once and then the FIFO is filled ahead of time to keep it from underflowing (or let it underflow and repeat the same cycle until stopped or changing the cycle).

My previous comment would be probably relevant in a scenario, where the exact number of cycles has to be output (some sort of audio output?), but then I think it probably does not makes a sense to change the period duration.

Well I am not sure at the end if we should do anything about this.

diegosueiro replied:
Perhaps we can leave as is (I used the same approach as the Linux kernel driver) for now and add a comment in the source code about this specific scenario that is not covered?

@MaureenHelm,
What do you think?

MaureenHelm replied:
I handled a similar case in pwm_mcux_ftm.c by logging a warning. You can do the same here for now.
https://github.com/zephyrproject-rtos/zephyr/blob/16ff8ca2c2aaade5b7658ccb594b9d52e1263e63/drivers/pwm/pwm_mcux_ftm.c#L61

MaureenHelm commented on drivers/pwm/pwm_imx.c:
```diff
@@ -0,0 +1,206 @@
+		return -EINVAL;
+	}
+
+	duty_cycle = 100 * pulse_cycles / period_cycles;
```
Another reason to move the formula is to avoid a compiler warning when you build with logging turned off.

MaureenHelm commented on boards/arm/colibri_imx7d_m4/Kconfig.defconfig:
```diff
@@ -58,5 +58,21 @@ config I2C_4
+	def_bool y
+
+config PWM_2
+	def_bool n
```
The `def_bool n` are redundant. I think all but PWM_1 should be removed. 

@ulfalizer Will you please confirm?

diegosueiro committed:
board: Add PWM support for colibri_imx7d_m4

Adds configurations and pinmux to support PWM on colibri_imx7d_m4

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
arch: Add i.MX7 PWM get clock frequency function

This patch adds the get_pwm_clock_freq function to be used by the
i.MX PWM driver.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

stanislav-poboril commented on drivers/pwm/pwm_imx.c:
```diff
@@ -0,0 +1,206 @@
+	} else {
+		PWM_PWMCR_REG(base) = PWM_PWMCR_SWR(1);
+		do {
+			k_sleep(1);
```
I didn't realize before that the k_busy_wait(500) is 500 microseconds, vs k_sleep(1) is one millisecond, but due to the default quantum is 10 milliseconds in Zephyr/ARM, it will sleep 10 milliseconds or more (unless kernel parameters are changed, you have maybe changed the system tick to 1 ms). Is this ok?

MaureenHelm commented on drivers/pwm/pwm_imx.c:
```diff
@@ -0,0 +1,206 @@
+			k_sleep(1);
+			cr = PWM_PWMCR_REG(base);
+		} while ((PWM_PWMCR_SWR(cr)) &&
+			 (wait_count++ < CONFIG_PWM_PWMSWR_LOOP));
```
This should be `++wait_count`

MaureenHelm commented on drivers/pwm/pwm_imx.c:
```diff
@@ -0,0 +1,206 @@
+			 (wait_count++ < CONFIG_PWM_PWMSWR_LOOP));
+
+		if (PWM_PWMCR_SWR(cr))
+			SYS_LOG_WRN("software reset timeout\n");
```
Need to wrap in braces

MaureenHelm reviewed (changes requested):
Just two small nits

diegosueiro committed:
drivers: add i.MX PWM driver

Adds the PWM shim driver for i.MX

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
